sensors/odometer.go

```diff
@@ -0,0 +1,183 @@
+}
+
+// TimedOdometerSensorReadingResponse represents an odometer sensor reading with a time & allows the caller
+// to know if the reading is from a replay movement sensor.
+type TimedOdometerSensorReadingResponse struct {
+	Position    *geo.Point
+	Orientation spatialmath.Orientation
+	ReadingTime time.Time
+	Replay      bool
+}
+
+// Odometer represents an Odometer movement sensor.
+type Odometer struct {
+	name            string
+	dataFrequencyHz int
+	Odometer        movementsensor.MovementSensor
+}
+
+// Name returns the name of the Odometer.
+func (odom Odometer) Name() string {
+	return odom.name
+}
+
+// DataFrequencyHz returns the data rate in ms of the Odometer.
+func (odom Odometer) DataFrequencyHz() int {
+	return odom.dataFrequencyHz
+}
+
+// TimedOdometerSensorReading returns data from the odometer movement sensor and the time the reading is from.
+func (odom Odometer) TimedOdometerSensorReading(ctx context.Context) (TimedOdometerSensorReadingResponse, error) {
+	replay := false
+
+	var readingTimePosition, readingTimeOrientation time.Time
+	var position *geo.Point
+	var orientation spatialmath.Orientation
+	var err error
+	for {
+		select {
+		case <-ctx.Done():
+			return TimedOdometerSensorReadingResponse{}, nil
+		default:
+			if readingTimePosition == defaultTime || readingTimePosition.Sub(readingTimeOrientation).Milliseconds() < 0 {
+				ctxWithMetadata, md := contextutils.ContextWithMetadata(ctx)
+				if position, _, err = odom.Odometer.Position(ctxWithMetadata, make(map[string]interface{})); err != nil {
+					return TimedOdometerSensorReadingResponse{}, errors.Wrap(err, "Position error")
+				}
+
+				readingTimePosition = time.Now().UTC()
+				if timeRequestedMetadata, ok := md[contextutils.TimeRequestedMetadataKey]; ok {
+					replay = true
+					if readingTimePosition, err = time.Parse(time.RFC3339Nano, timeRequestedMetadata[0]); err != nil {
+						return TimedOdometerSensorReadingResponse{}, errors.Wrap(err, replayTimestampErrorMessage)
+					}
+				}
+			}
+
+			if readingTimeOrientation == defaultTime || readingTimeOrientation.Sub(readingTimePosition).Milliseconds() < 0 {
+				ctxWithMetadata, md := contextutils.ContextWithMetadata(ctx)
+				if orientation, err = odom.Odometer.Orientation(ctxWithMetadata, make(map[string]interface{})); err != nil {
+					return TimedOdometerSensorReadingResponse{}, errors.Wrap(err, "Orientation error")
+				}
+
+				readingTimeOrientation = time.Now().UTC()
+				if timeRequestedMetadata, ok := md[contextutils.TimeRequestedMetadataKey]; ok {
+					replay = true
+					if readingTimeOrientation, err = time.Parse(time.RFC3339Nano, timeRequestedMetadata[0]); err != nil {
+						return TimedOdometerSensorReadingResponse{}, errors.Wrap(err, replayTimestampErrorMessage)
+					}
+				}
+			}
+			if math.Abs(float64(readingTimeOrientation.Sub(readingTimePosition).Milliseconds())) < replayTimeToleranceMsec {
```

Comment:
Can you make sure `TimedOdometerSensorReading` is always called with a timeout, in case this condition is never satisfied?